src/ui_nbgl.c

```diff
@@ -0,0 +1,433 @@
+ *   (c) 2022 Ledger
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ ********************************************************************************/
+
+#ifdef HAVE_NBGL
+#include "ui_nbgl.h"
+#include "btchip_display_variables.h"
+#include "btchip_internal.h"
+
+#include "btchip_bagl_extensions.h"
+
+enum { MESSAGE_TYPE, TRANSACTION_TYPE };
+
+static nbgl_page_t *pageContext;
+static UiContext_t uiContext = {.transaction_prompt_done = 0};
+
+// User action callbacks
+static void approved_user_action_callback(void) {
+  if (!btchip_bagl_user_action(1)) {
+    ui_idle_flow();
+  }
+}
+
+static void approved_user_action_processing_callback(void) {
+  if (!btchip_bagl_user_action(1)) {
+    nbgl_useCaseSpinner("Processing");
+  }
+}
+
+static void abandon_user_action_callback(void) {
+  if (!btchip_bagl_user_action(0)) {
+    ui_idle_flow();
+  }
+}
+
+static void approved_user_action_message_signing_callback(void) {
+  btchip_bagl_user_action_message_signing(1);
+  ui_idle_flow();
+}
+
+static void abandon_user_action_message_signing_callback(void) {
+  btchip_bagl_user_action_message_signing(0);
+  ui_idle_flow();
+}
+
+static void approved_user_action_display_processing_callback(void) {
+  btchip_bagl_user_action_display(1);
+  nbgl_useCaseSpinner("Processing");
+}
+
+static void approved_user_action_display_callback(void) {
+  btchip_bagl_user_action_display(1);
+  ui_idle_flow();
+}
+
+static void abandon_user_action_display_callback(void) {
+  btchip_bagl_user_action_display(0);
+  ui_idle_flow();
+}
+
+static void approved_user_action_signtx_callback(void) {
+  btchip_bagl_user_action_signtx(1, 0);
+  ui_idle_flow();
+}
+
+static void abandon_user_action_signtx_callback(void) {
+  btchip_bagl_user_action_signtx(0, 0);
+  ui_idle_flow();
+}
+
+static void releaseContext(void) {
+  if (pageContext != NULL) {
+    nbgl_pageRelease(pageContext);
+    pageContext = NULL;
+  }
+}
+
+// Status
+static void abandon_status(void) {
+  nbgl_useCaseStatus(uiContext.abandon_status, false, uiContext.abandon_cb);
+  uiContext.transaction_prompt_done = false;
+}
+
+static void approved_status(void) {
+  nbgl_useCaseStatus(uiContext.approved_status, true, uiContext.approved_cb);
+  uiContext.transaction_prompt_done = false;
+}
+
+static void status_callback(bool confirm) {
+  if (confirm) {
+    approved_status();
+  } else {
+    abandon_status();
+  }
+}
+
+// Prompt Cancel
+static void prompt_cancel(uint8_t type) {
```

Comment:
You might want to define a type for your enum and update the prompt_cancel prototype, for instance : 

`static void prompt_cancel(PromptType prompt_type)`